### ISSUE TITLE
Respect GHC version specified in cabal.project

### ISSUE THREAD

**Issue by AlexeyRaga** [haskell/haskell-ide-engine#1023]
In my project I have the following `cabal.project` file:

```
with-compiler: ghc-8.6.3
packages: .

```

However, as far as I can tell, HIE doesn't use GHC 8.6.3 and tries to use the "default" one (which is 8.4.4 in my case):

```
readCreateProcess: /Users/araga/src/github/hie/.stack-work/install/x86_64-osx/lts-12.26/8.4.4/bin/cabal-helper-wrapper "--with-ghc=ghc" "--with-ghc-pkg=ghc-pkg" "--with-cabal=cabal" "v1-style" "/Users/araga/src/haskell-works/hw-xml" "/Users/araga/src/haskell-works/hw-xml/dist-newstyle/build/x86_64-osx/ghc-8.4.4/hw-xml-0.1.0.4" "package-db-stack" "flags" "compiler-version" "ghc-merged-pkg-options" "config-flags" "non-default-config-flags" "ghc-src-options" "ghc-pkg-options" "ghc-lang-options" "ghc-options" "source-dirs" "entrypoints" "needs-build-output" (exit 1): failed
```

Executing this command manually I see that it tries to load a config file for GHC 8.4.4:

```
cabal-helper-wrapper: /Users/araga/src/haskell-works/hw-xml/dist-newstyle/build/x86_64-osx/ghc-8.4.4/hw-xml-0.1.0.4/setup-config: openFile: does not exist (No such file or directory)
```

Expected behaviour: use GHC that is specified in `cabal.project`.

**Comment by AlexeyRaga** [comment]
Perhaps the difference is:

```
$ cabal new-exec ghc -- --numeric-version
8.6.3
```

vs.

```
$ cabal exec ghc -- --numeric-version
8.4.4
```

**Comment by AlexeyRaga** [comment]
Indeed
```
$ hie-wrapper
2019-01-04 23:52:40.826236 [ThreadId 4] - run entered for hie-wrapper(hie-wrapper) Version 0.4.0.1, Git revision 0c70fb9521f70c2f98fd137abd7c90ddb356b104 (dirty) (2306 commits) x86_64 ghc-8.6.1
2019-01-04 23:52:40.828297 [ThreadId 4] - Current directory:/Users/araga/src/haskell-works/hw-xml
2019-01-04 23:52:41.198228 [ThreadId 4] - Cradle directory:/Users/araga/src/haskell-works/hw-xml
2019-01-04 23:52:41.199652 [ThreadId 4] - Using plain GHC version
2019-01-04 23:52:41.266881 [ThreadId 4] - Project GHC version:8.4.4
```

**Comment by AlexeyRaga** [comment]
This line seems to be responsible:
https://github.com/haskell/haskell-ide-engine/blob/master/src/Haskell/Ide/Engine/Plugin/Base.hs#L125

Shouldn't it be calling to `cabal v2-exec ghc -- --numeric-version` instead?

**Comment by jrp2014** [comment]
I have been trying to get hie / hie-wrapper to work with VSCode on Ubuntu 18.10.

The build process is fairly painless.  I tried first with 'cabal new-install exe:hie' and then with 'make build-all'.  The latter takes for ever and installs gigabytes of stuff as it installs multiple versions of GHC and libraries and rebuilds lots of libraries. etc.

Perhaps this is just a documentation thing, but although the Haskell syntax checker seems to be in operation in VSCode, the Haskell Language Server extension doesn't appear to do much, and it's not clear to me how to check what might be going wrong.

**Comment by AlexeyRaga** [comment]
@jrp2014 What do you mean by "doesn't appear to do much"? Which bits of that "much" it does or doesn't do?
Also, what `Output -> Haskell HIE` says? Usually, if there is an error you'd see the line there.

If it is related to this bug, then yes, it is still present (unless resolved last night) and my workaround, for now, to switch to the required GHC when working on a project.

**Comment by jrp2014** [comment]
Anything.  I don't even see an Output for HIE (or any output, or debugging).  I should have been clearer, but I do get syntax highlighting and if I hover over an identifier it highlights its occurrences, but not types, and there are no complaints if I introduce a syntax error.

**Comment by AlexeyRaga** [comment]
If you don't even have Output for HIE (like the one on my screenshot) then I'd suggest you check if the extension is even installed/enabled, and perhaps other logs that are there (Extensions, Main, etc.).

![image](https://user-images.githubusercontent.com/1394391/51310603-1ff65080-1a9b-11e9-80dc-a507c6954de3.png)

At this point, if the VSCode extension isn't even starting, I'd suggest that the problem may not be with HIE, but with the extension itself so consider raising an issue here (https://github.com/alanz/vscode-hie-server) and providing more info (whether the extension is installed, which version, logs from VSCode trying to start the extension)

**Comment by AlexeyRaga** [comment]
@jrp2014  BTW, are you sure that your problem is related to this my issue of not respecting GHC version in `cabal.project`? What happens if you remove `with-compiler` from `cabal.project` or remove `cabal.project` at all? Does it start then?
Because so far _it looks like_ your problem is different, like VSCode even doesn't attempt running HIE or something like it.

**Comment by power-fungus** [comment]
I had similar behaviour when ghc-mod is unable to find cabal-helper-wrapper. This is visible in stderr of hie. I do not know how to get that in vsc

**Comment by bubba** [comment]
Stderr is also shown in the output window of vscode

**Comment by jrp2014** [comment]
Is a Cabal file a requirement of getting any life out of the extension?  (It appears to be loaded, with the Syntax extension, when I look at the VSCode extensions.)  I have tried just loading a stand-alone .hs file to simplify things.

**Comment by bubba** [comment]
HIE does work in standalone .hs files with other clients, but for some reason on VS code it doesn’t kick it off until there’s a cabal file (I believe).

**Comment by jrp2014** [comment]
Thanks,  It would be helpful to document the limitations, and the required dependencies, in more detail.

So far, on MacOS, VSCode, and stand-alone `.hs` files HIE, seems to fire up OK.  But then when I try to peek at a definition it goes off and does a lot of stuff in the background culminating in `readCreateProcess: stack "build" "--only-configure" "." (exit 1): failed`.  (This is not surprising as there is no stack file, but it isn't a great experience.). Similarly if I put a couple of separate `.hs` files in a folder it thinks that they form a project of some sort and complains about there being more than one definition of `main`.  (Some of them are linked with explicit imports and others not.)

On Ubuntu VSCode, standalone `.hs` files don't trigger HIE, so far as I can see.  Files that are part of a .cabal project seem to trigger a `readCreateProcess` failure.

Perhaps I'll have better luck with HIE from ALE / Vim.

**Comment by AlexeyRaga** [comment]
I think I saw that `readCreateProcess` too.
But the fact that it tries to call to `stack` is strange since you have `.cabal` file and no `stack.yaml` file.

I am also on MacOS + VSCode, but when I see `readCreateProcess` it usually because of https://github.com/haskell/haskell-ide-engine/issues/1015 so the suggested workaround applies.

Also, AFAIK HIE tries to look inside `dist-newstyle` so before starting HIE on the project it helps to build it with `cabal new-build` and to apply a workaround from https://github.com/haskell/haskell-ide-engine/issues/1015

```
for x in $(find dist-newstyle -name setup-config | grep '/opt/setup-config$' | sed 's|/opt/setup-config$||g'); do
  ( cd $x
    ln -fs opt/setup-config setup-config
  )
done
``` 

I know that it is not ideal, but if you are hitting the same issue then it may help :)

**Comment by teto** [comment]
thanks for the tip. As a matter of fact, setup-config was in yet another folder here : 
dist-newstyle/build/x86_64-linux/ghc-8.6.3/netlink-pm-1.0.0/x/daemon/setup-config (when hie was expecting dist-newstyle/build/x86_64-linux/ghc-8.6.3/netlink-pm-1.0.0/setup-config)

**Comment by AlexeyRaga** [comment]
Since the milestone `2019-03` has passed, do we have any hope for better supporting `cabal new-build` and `cabal.project` any time soon?

**Comment by fendor** [comment]
There is active work going at the moment, including adding support upstream

**Comment by alanz** [comment]
I think this issue will be properly addressed when the hie-bios is fully integrated.

**Comment by fendor** [comment]
We can address the issue about honouring the right ghc from now on! 
Closing this in favour of #943
